### PR TITLE
fix #630 Reduce WorkQueueProcessor/TopicProcessor sink serialization

### LINKED ISSUE #630: WorkQueueProcessor.share().sink() FluxSink implementation should not serialize requests
`BlockingSink` in 3.0.x did not serialize and could be used for back-pressure aware emission from multiple producer threads onto ring-buffer thus demultiplexing incoming events.

The performance impact of using `.sink/FluxSink` which serializes is significant.

The use case where this is important is when you have multiple producer threads that cannot respond to `onRequest` such as socket based event sources.  In these cases you need multi-threaded producers with configurable overflowStrategy. WAIT, DROP (with Consumer) etc. but without overhead of unnecessary serialization.

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

File: reactor-core/src/test/java/reactor/core/publisher/TopicProcessorTest.java [hunk 1]
```diff
@@ -834,4 +836,46 @@ private void assertProcessor(TopicProcessor<Integer> processor,
...
 			assertEquals(requestTaskExecutor, processor.requestTaskExecutor);
 	}
+
+	@Test
+	public void serializedSinkSingleProducer() throws Exception {
+		TopicProcessor<Integer> processor = TopicProcessor.<Integer>builder()
+				.share(false)
+				.build();
+		FluxSink<Integer> sink = processor.sink();
+		assertTrue("Should not be serialized", sink instanceof SerializedSink);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/TopicProcessorTest.java]: message should be "Should be serialized". Alternatively, use AssertJ: `Assertions.assertThat(sink).isInstanceOf(SerializedSink.class)`

File: reactor-core/src/test/java/reactor/core/publisher/TopicProcessorTest.java [hunk 2]
```diff
@@ -834,4 +836,46 @@ private void assertProcessor(TopicProcessor<Integer> processor,
...
+		sink = sink.next(1);
+		assertTrue("Should not be serialized", sink instanceof SerializedSink);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/TopicProcessorTest.java]: invert message here too (see above)

File: reactor-core/src/test/java/reactor/core/publisher/TopicProcessorTest.java [hunk 3]
```diff
@@ -834,4 +836,46 @@ private void assertProcessor(TopicProcessor<Integer> processor,
...
+		sink = sink.onRequest(n -> {});
+		assertTrue("Should not be serialized", sink instanceof SerializedSink);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/TopicProcessorTest.java]: invert message here too (see above)

File: reactor-core/src/test/java/reactor/core/publisher/WorkQueueProcessorTest.java [hunk 4]
```diff
@@ -1427,4 +1428,111 @@ private void assertProcessor(WorkQueueProcessor<Integer> processor,
...
 			assertEquals(requestTaskExecutor, processor.requestTaskExecutor);
 	}
+
+	@Test
+	public void serializedSinkSingleProducer() throws Exception {
+		WorkQueueProcessor<Integer> queueProcessor = WorkQueueProcessor.<Integer>builder()
+				.share(false).build();
+
+		FluxSink<Integer> sink = queueProcessor.sink();
+		assertTrue("Should not be serialized", sink instanceof SerializedSink);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/WorkQueueProcessorTest.java]: invert message here too (see above)

File: reactor-core/src/test/java/reactor/core/publisher/WorkQueueProcessorTest.java [hunk 5]
```diff
@@ -1427,4 +1428,111 @@ private void assertProcessor(WorkQueueProcessor<Integer> processor,
...
+		sink = sink.next(1);
+		assertTrue("Should not be serialized", sink instanceof SerializedSink);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/WorkQueueProcessorTest.java]: invert message here too (see above)

File: reactor-core/src/test/java/reactor/core/publisher/WorkQueueProcessorTest.java [hunk 6]
```diff
@@ -1427,4 +1428,111 @@ private void assertProcessor(WorkQueueProcessor<Integer> processor,
...
+		sink = sink.onRequest(n -> {
+		});
+		assertTrue("Should not be serialized", sink instanceof SerializedSink);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/WorkQueueProcessorTest.java]: invert message here too (see above)

**Comment by rajinisivaram** [comment]
@simonbasle Thank you for the review. Have fixed the assertions.

**Comment by smaldini** [comment]
Not a bad idea, just add an extra line of stack and probably a volatile boolean.
We should be able to switch to the onRequest based serialization after the return of the sink callback (we cannot declare onRequest later than that).

**Pull request merged**